src/txrebroadcast.cpp

```diff
@@ -0,0 +1,273 @@
+// Copyright (c) 2021 The Bitcoin Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#include <chainparams.h>
+#include <consensus/consensus.h>
+#include <logging.h>
+#include <miner.h>
+#include <node/blockstorage.h>
+#include <script/script.h>
+#include <txrebroadcast.h>
+#include <util/time.h>
+
+#include <boost/multi_index/hashed_index.hpp>
+#include <boost/multi_index/member.hpp>
+#include <boost/multi_index/ordered_index.hpp>
+#include <boost/multi_index_container.hpp>
+
+/** We rebroadcast upto 3/4 of max block weight to reduce noise due to
+ * circumstances such as miners mining priority transactions. */
+static constexpr float REBROADCAST_WEIGHT_RATIO{0.75};
+
+/** Default minimum age for a transaction to be rebroadcast */
+static constexpr std::chrono::minutes REBROADCAST_MIN_TX_AGE{30min};
+
+/** Maximum number of times we will rebroadcast a transaction */
+static constexpr int MAX_REBROADCAST_COUNT{6};
+
+/** Minimum amount of time between returning the same transaction for
+ * rebroadcast */
+static constexpr std::chrono::hours MIN_REATTEMPT_INTERVAL{4h};
+
+/** The maximum number of entries permitted in m_attempt_tracker */
+static constexpr int MAX_ENTRIES{500};
+
+/** The maximum age of an entry ~3 months */
+static constexpr std::chrono::hours MAX_ENTRY_AGE{24h * 30 * 3};
+
+struct RebroadcastEntry {
+    RebroadcastEntry(std::chrono::microseconds now_time, uint256 wtxid)
+        : m_last_attempt(now_time),
+          m_wtxid(wtxid) {}
+
+    std::chrono::microseconds m_last_attempt;
+    const uint256 m_wtxid;
+    int m_count{1};
+};
+
+/** Used for multi_index tag  */
+struct index_by_last_attempt {};
+
+class indexed_rebroadcast_set : public
+boost::multi_index_container<
+    RebroadcastEntry,
+    boost::multi_index::indexed_by<
+        // sorted by wtxid
+        boost::multi_index::hashed_unique<
+            boost::multi_index::tag<index_by_wtxid>,
+            boost::multi_index::member<RebroadcastEntry, const uint256, &RebroadcastEntry::m_wtxid>,
+            SaltedTxidHasher
+        >,
+        // sorted by last rebroadcast time
+        boost::multi_index::ordered_non_unique<
+            boost::multi_index::tag<index_by_last_attempt>,
+            boost::multi_index::member<RebroadcastEntry, std::chrono::microseconds, &RebroadcastEntry::m_last_attempt>
+        >
+    >
+>{};
+
+TxRebroadcastHandler::~TxRebroadcastHandler() = default;
+
+TxRebroadcastHandler::TxRebroadcastHandler(const CTxMemPool& mempool, const ChainstateManager& chainman, const CChainParams& chainparams)
+    : m_mempool{mempool},
+      m_chainman{chainman},
+      m_chainparams(chainparams),
+      m_attempt_tracker{std::make_unique<indexed_rebroadcast_set>()},
+      m_max_filter(1500, 0.0001){}
```

Comment:
This gives a 1-in-10k fp rate, so once the bloom filter has had time to fill up, we should expect each node to incorrectly not rebroadcast 1-in-10k transactions. However, the nTweak parameter should ensure that this is affects different transactions on different nodes, so I think this should be fine.